File: packages/js/core/src/algorithms/uri-resolution/UriResolutionHistory.ts
import { Uri } from "../..";

export type UriResolutionStack = { 
  resolver: string;
  sourceUri: Uri;
  result: {
    uri: Uri;
    api: boolean;
  }
}[];

export class UriResolutionHistory {
  constructor(public readonly stack: UriResolutionStack) {
  }

  getResolvers(): string[] {
    return this.stack.map(({ resolver }) => resolver);
  }

  getUris(): Uri[] {
    return this.stack.map(({ result: { uri } }) => uri);
  };

  //Resolution path includes the list of resolvers that redirected to another URI or returned the API
  getResolutionPath(): UriResolutionHistory {
    const path: UriResolutionHistory = new UriResolutionHistory(
      this.stack.filter(x => x.sourceUri.uri !== x.result.uri.uri || x.result.api)
    );

    return path;
  }
};

File: packages/py/core/src/uri_resolution/core/types/uri_resolution_history.py
from typing import List


class UriResolutionHistory:
    def __init__(self, stack: "UriResolutionStack"):
        self.stack = stack

    @property
    def uri_resolvers(self) -> List[str]:
        return [x.uri_resolver for x in self.stack]

    @property
    def uris(self) -> List["Uri"]:
        return list(set(x.result.uri for x in self.stack))

    def get_resolution_path(self) -> "UriResolutionHistory":
        path = UriResolutionHistory([x for x in self.stack if x.source_uri.uri != x.result.uri.uri or x.result.api])
        return path
